text/0000-gen-fn.md

```diff
@@ -0,0 +1,429 @@
+
+## different syntax for `#[rustc_gen] fn`:
+
+This RFC explicitly picks an attribute, as that has no conflicts with any other syntax, even within macros, and
+does not pick any option that may influence how experimental users think about syntax.
+There are many options to choose from, and we'll have to decide on one before stabilization.
+Some options are:
+
+```rust
+fn foo(args) yield item
+fn foo(args) yields item
+fn foo(args) => item
+fn* foo(args) -> item // or any of the `fn foo` variants for the item type
+gen fn foo(args) // or any of the above variants for the item type
+gen foo(args) // or any of the above variants for the item type
+generator fn foo(args) // or any of the above variants for the item type
+```
+
+The design space here is very large, but either way, I propose to reserve the `gen` keyword.
+
+# Prior art
+[prior-art]: #prior-art
+
+## Python
+
+Python has equivalent functionality to `#[rustc_gen] fn`: any function that uses `yield` internally.
+The main difference is that raising an exception automatically passes the exception outwards, instead of yielding an `Err()` element.
+
+```python
+def odd_dup(values):
+    for value in values:
+        if is_odd(value):
+            yield value * 2
+```
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Keyword
+
+Should we use `iter` as the keyword, as we're producing `Iterator`s?
+We could use `gen` as proposed in this RFC and later extend its abilities to more powerful generators.
+
+[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=efeacb803158c2ebd57d43b4e606c0b5)
+
+```rust
+#![feature(generators)]
+#![feature(iter_from_generator)]
+
+fn main() {
+    let mut it = std::iter::from_generator(|| {
+        yield 1
+    });
+
+    assert_eq!(it.next(), Some(1));
+    assert_eq!(it.next(), None);
+    it.next(); // panics
+}
+```
+
+## Panicking
+
+What happens when `Iterator::next` is called again on a `#[rustc_gen]` block that panicked? Do we need to poison the iterator?
+
+## Contextual keyword
+
+Popular crates (like `rand`) have methods called [`gen`][Rng::gen]. If we forbid those, we are forcing those crates to make a major version bump when they update their edition, and we are requiring any users of those crates to use `r#gen` instead of `gen` when calling that method.
+
+We could choose to use a contextual keyword and only forbid `gen` in
+
+* bindings,
+* field names (due to destructuring bindings),
+* enum variants,
+* and type names
+
+This should avoid any parsing issues around `gen` followed by `{` in expressions.
+
+[Rng::gen]: https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen
+
+## `Iterator::size_hint`
+
+Should we try to compute a conservative `size_hint`? This will reveal information from the body of a generator,
+but at least for simple cases users will likely expect `size_hint` to not just be the default.
```

Comment:
A manual mechanism to specify `size_hint` can be added later, but also it would be possible to automatically calculate it in rustc if every looped iterator has a `size_hint` set: single `yield` statements increment the counter, every `for _ in iter` adds `iter.size_hint()` to it, and nested iterators are multiplied. This breaks down as soon as there's a single iterator without a valid `size_hint` though, which is why the manual mechanism should also exist. All of this is *not* needed for an MCP though.